substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[frame_support::pallet]
+pub mod pallet {
+  use scale::{Encode, Decode};
+  use scale_info::TypeInfo;
+
+  use frame_system::pallet_prelude::*;
+  use frame_support::pallet_prelude::*;
+
+  use validator_sets_primitives::*;
+
+  #[pallet::config]
+  pub trait Config: frame_system::Config + TypeInfo {
+    type RuntimeEvent: IsType<<Self as frame_system::Config>::RuntimeEvent> + From<Event<Self>>;
+  }
+
+  #[pallet::genesis_config]
+  #[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen)]
+  pub struct GenesisConfig<T: Config> {
+    /// Bond requirement to join the initial validator set.
+    /// Every participant at genesis will automatically be assumed to have this much bond.
+    /// This bond cannot be withdrawn however as there's no stake behind it.
+    pub bond: Amount,
+    /// Amount of coins to spawn the network with in the initial validator set.
+    pub coins: Coin,
+    /// List of participants to place in the genesis set.
+    pub participants: Vec<T::AccountId>,
+  }
+
+  #[cfg(feature = "std")]
+  impl<T: Config> Default for GenesisConfig<T> {
+    fn default() -> Self {
+      GenesisConfig { bond: Amount(0), coins: Coin(0), participants: vec![] }
+    }
+  }
+
+  // Max of 16 coins per validator set
+  // At launch, we'll have BTC, ETH, DAI, and XMR
+  // In the future, these will be split into separate validator sets, so we're already not
+  // planning expansion beyond just a few coins per validator set
+  // The only case which really makes sense for multiple coins in a validator set is:
+  // 1) The coins are small, easy to run, and make no sense to be in their own set
+  // In this case, it's still hard to ask validators to run 16 different nodes
+  // 2) The coins are all on the same network yet there's no DEX on-chain
+  // In these cases, it'd be hard to find and justify 16 different coins from that single chain
+  // This could probably be just 8, yet 16 is a hedge for the unforseen
+  // If necessary, this can be increased with a fork
+  type MaxCoinsPerSet = ConstU32<16>;
```

Comment:
It has to be a ConstU32 as there's no const usize, and u32 takes the place of usize. It's 16 because I never see a single set managing more than 8 and 16 was a sane buffer.

Even if we expand with 4 more blockchains before multiple sets, we'd be at 8. If we ever list XHV (not an endorsement, they just don't have a on-chain DEX and would rely on a group such as ourselves), they technically have 11 yet only 3 or 4 worth a damn.

We can also trivially increase this in the future (with an upgrade, without migration logic).